### PR TITLE
[Gravatar] - Add Gravatar info UI in user profile settings

### LINKED ISSUE #22551: Task: Add informative Gravatar UI to the profile editing
As part of the recent Gravatar efforts we are adding an informative card to the profile editing. Because currently users don't know that their profile info is linked to their Gravatar profile and making changes here also makes changes there. This is also a way to introduce Gravatar to users.

Screenshot 2024-02-06 at 11 41 24

### PR DESCRIPTION
**Pull request by pinarol**
Fixes https://github.com/wordpress-mobile/WordPress-iOS/issues/22551

We are looking into how we can make our users aware that when editing a profile, they are actually editing their Gravatar profile. This PR adds some informative UI for that purpose.

**Jetpack:**

<table>
  <tr>
    <td></td>
    <td></td>
  </tr>
</table>

**WordPress:**

<table>
  <tr>
    <td></td>
    <td></td>
  </tr>
</table>

To test:

Try on both JP and WP:

- Go to the profile tab > My Profile
- Observe: There's a new informative UI as demonstrated above ☝️ 
- Try updating those 4 fields
- Observe you see a notice message saying "Updates might take some time to sync with your Gravatar profile."
 
(There's a possibility that we use a different copy for the notice message. But it should not be a blocker for reviewing the PR. Just letting you know that I might push a new commit for this.)

## Regression Notes

1. Potential unintended areas of impact

Account settings. App settings. This time the notice should NOT show upon updating fields.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

I tried changing some fields in Account settings. App settings. Verified that the notice in NOT shown in that case.

3. What automated tests I added (or what prevented me from doing so)

I looked into writing some snapshot tests but this repo doesn't have any of those. That requires a new snapshot library added to the repo and that would be out of context of this PR.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [x] Portrait and landscape orientations.
- [x] Light and dark modes.
- [x] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WordPress/Classes/ViewRelated/Tools/SettingsCommon.swift [hunk 1]
```diff
@@ -75,11 +79,14 @@ extension SettingsController {
 
         controller.title = title
         controller.onValueChanged = {
-            value in
-
-            let change = changeType(value)
+            newValue in
+            guard value != newValue else { return }
```
Comment (pinarol) [inline WordPress/Classes/ViewRelated/Tools/SettingsCommon.swift:83]: I added this guard here because SettingsMultiTextViewController calls `onValueChanged` whether the value changes or not. And this causes the notice to appear even if you haven't changed the "About Me" field. This is different from `SettingsTextViewController` which only calls `onValueChanged` if the value has really been changed.

Ideally`SettingsMultiTextViewController` should act as `SettingsTextViewController`, but changing its current behavior would break other places that already use it. 🤷‍♀️

File: WordPress/WordPress.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -19216,13 +19224,13 @@
...
 				sk,
 			);
-			mainGroup = 29B97314FDCFA39411CA2CEA;
+			mainGroup = 29B97314FDCFA39411CA2CEA /* CustomTemplate */;
 			packageReferences = (
 				3FF1442E266F3C2400138163 /* XCRemoteSwiftPackageReference "ScreenObject" */,
 				3FC2C33B26C4CF0A00C6D98F /* XCRemoteSwiftPackageReference "XCUITestHelpers" */,
 				17A8858B2757B97F0071FCA3 /* XCRemoteSwiftPackageReference "AutomatticAbout-swift" */,
 				3F3B23C02858A1B300CACE60 /* XCRemoteSwiftPackageReference "test-collector-swift" */,
-				3F411B6D28987E3F002513AE /* XCRemoteSwiftPackageReference "lottie-ios.git" */,
```
Comment (pinarol) [inline WordPress/WordPress.xcodeproj/project.pbxproj]: I am not sure what I did to cause these changes 🤔. I'll revert those.

**Comment by pinarol** [comment]
Okay I think this is ready to review.

**Review by etoledom — APPROVED** [review]
Looks quite good and tidy up 🎉 

The UI also looks nice! 

- **I found one detail which we probably should attend before merging:**

When the user changes the text size on iOS settings, the Info cell's text size won't change dynamically. It requires a restart of the app for the change to take effect.




- **One more detail worth mentioning, but it's outside of the scope of this PR:**

The link icon on RTL languages should be mirrored:



I say it's outside of the scope of this PR because there are other instances of this same problem in the app: 



- **Note:** We are also using a different logo design for WP and JetPack 🤔  
Not sure if it's worth coordinating these designs somehow:

CleanShot 2024-02-09 at 18 27 34@2x

But I'd say this can also be discussed after this PR is merged.

--- 

In my opinion, the only thing to be atended is the dynamic text size change mentioned before, and we can move forward 👍

**Comment by pinarol** [comment]
Thanks for the review.

> When the user changes the text size on iOS settings, the Info cell's text size won't change dynamically. It requires a restart of the app for the change to take effect.

https://github.com/wordpress-mobile/WordPress-iOS/pull/22563/commits/5e2d91495ddeadc0a8205fd88eb32265c2ae9013 I did necessary changes to address this in GravatarRowInfo. But the external gravatar link is left out because it's an attributed string and it uses `TextStyle.caption.uiFont` from `UILabel+DesignSystem`. 

In order for it to adapt font size changes dynamically we need to change the UIFont extension in UILabel+DesignSystem.

so instead of this:

> static let caption = DynamicFontHelper.fontForTextStyle(.caption1, fontWeight: .regular)

we need something like this:

> static var caption: UIFont { DynamicFontHelper.fontForTextStyle(.caption1, fontWeight: .regular) }

There's an ongoing design system effort and @alpavanoglu is leading it. As far as I understand they are still working on it. So I'll leave this decision to them. I am not sure to what degree they aim to support dynamic fonts. And changing that file would exceed the context of this PR. 

> Note: We are also using a different logo design for WP and JetPack 🤔

These logos are same with the ones in the splash screen. They have extra paddings around the logo and it matches with the Gravatar logo perfectly. The other ones don't look good when put side by side with the Gravatar logo. (BTW I had the 👍 from our designer about how the new UI looks in WP and JP) 

<image src="https://github.com/wordpress-mobile/WordPress-iOS/assets/5032900/acf9f40f-1943-47c3-9f76-7f39d26fcf5c" width=250/>

**Comment by alpavanoglu** [comment]
@pinarol 
> As far as I understand they are still working on it.

We are. However I wasn't aware that this problem existed on AttributedString. I'll update the values as computed vars from stored constants.

File: WordPress/Classes/ViewRelated/Me/My Profile/Gravatar/GravatarInfoRow.swift [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        label.text = Constants.infoText
+        label.adjustsFontForContentSizeCategory = true
+        label.numberOfLines = 0
+        return label
+    }()
+
+    private var linkTextView: UITextView = {
+        let text = UITextView()
+        text.translatesAutoresizingMaskIntoConstraints = false
+        text.isUserInteractionEnabled = true
```
Comment (alpavanoglu) [inline WordPress/Classes/ViewRelated/Me/My Profile/Gravatar/GravatarInfoRow.swift]: Nit: I think the default value of `isUserInteractionEnabled` is true so this line might be redundant.

File: WordPress/Classes/ViewRelated/Me/My Profile/Gravatar/GravatarInterceptingLogosView.swift [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    private var appLogo: UIImage? {
+        return AppConfiguration.isJetpack ? Constants.jetpackLogo : Constants.wordpressLogo
+    }
+
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+        addSubview(appLogoImageView)
+        addSubview(gravatarLogoImageView)
+        gravatarLogoImageView.layer.zPosition = 1
```
Comment (alpavanoglu) [inline WordPress/Classes/ViewRelated/Me/My Profile/Gravatar/GravatarInterceptingLogosView.swift]: Why was this necessary? Would `appLogoImageView` have a zPosition 0 and gravatarLogoImageView 1 by default due to the order we add them to the view hierarchy?

**Review by alpavanoglu — APPROVED** [review]
LGTM 👍 
I dropped 2 minor comments.

File: WordPress/Classes/ViewRelated/Me/My Profile/Gravatar/GravatarInterceptingLogosView.swift (hunk 4 above)
Comment (pinarol) [reply WordPress/Classes/ViewRelated/Me/My Profile/Gravatar/GravatarInterceptingLogosView.swift]: According to designs the Gravatar logo should appear on top(check out the intersecting logos), but I think we don't need to set the zPosition for that. Let me revisit this.

Comment (pinarol) [reply WordPress/Classes/ViewRelated/Me/My Profile/Gravatar/GravatarInterceptingLogosView.swift]: Replaced with bringSubviewToFront.

File: WordPress/Classes/ViewRelated/Me/My Profile/Gravatar/GravatarInfoRow.swift (hunk 3 above)
Comment (pinarol) [reply WordPress/Classes/ViewRelated/Me/My Profile/Gravatar/GravatarInfoRow.swift]: Correct. This was a label so apparently I forgot to remove that after turning it to UITextView. Done now.

**Comment by etoledom** [comment]
> I did necessary changes to address this in GravatarRowInfo.

Thank you! Tested and working great 🎉 

> But the external gravatar link is left out because it's an attributed string and it uses TextStyle.caption.uiFont from UILabel+DesignSystem.

Ok, let's go with it. Thanks for the clarification 🙏

**Comment by Luchadores** [comment]
These look good to me as well! But let's keep in mind we might have to adjust the copy at some point, as it's being discussed here: https://gravatarp2.wordpress.com/2024/02/08/mobile-editing-i4/#comment-1156

**Comment by pinarol** [comment]
Thanks for the ping @Luchadores. I am planning to hide the changes until you are certain about the copy. But we have time until end of Friday (I mean until the release cut). If you come to an agreement before that I won't need to hide but just update the UI.
